File: src/main/java/nl/chimpgamer/networkmanager/extensions/discordbot/commands/discord/PlaytimeCommand.java
package nl.chimpgamer.networkmanager.extensions.discordbot.commands.discord;

import com.jagrosh.jdautilities.command.Command;
import com.jagrosh.jdautilities.command.CommandEvent;
import net.dv8tion.jda.core.Permission;
import net.dv8tion.jda.core.entities.ChannelType;
import net.dv8tion.jda.core.entities.MessageEmbed;
import nl.chimpgamer.networkmanager.api.models.player.Player;
import nl.chimpgamer.networkmanager.api.utils.TimeUtils;
import nl.chimpgamer.networkmanager.extensions.discordbot.DiscordBot;
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.CommandSetting;
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.DCMessage;
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.JsonEmbedBuilder;
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.Utils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.LinkedList;
import java.util.List;
import java.util.UUID;

public class PlaytimeCommand extends Command {
    private final DiscordBot discordBot;

    public PlaytimeCommand(DiscordBot discordBot) {
        this.discordBot = discordBot;
        this.name = CommandSetting.DISCORD_PLAYTIME_COMMAND.getAsString();
        this.cooldown = 3;
        this.botPermissions = new Permission[]{Permission.MESSAGE_WRITE};
        this.guildOnly = true;
    }

    @Override
    protected void execute(CommandEvent event) {
        if (!event.isFromType(ChannelType.TEXT)) {
            return;
        }
        if (!CommandSetting.DISCORD_PLAYTIME_ENABLED.getAsBoolean()) {
            return;
        }

        if (event.getArgs().isEmpty()) {
            UUID uuid = this.getDiscordBot().getDiscordUserManager().getUuidByDiscordId(event.getAuthor().getId());
            if (uuid == null) {
                this.getDiscordBot().getLogger().warning(event.getAuthor().getName() + " tried to use the playtime command but is not registed!");
                return;
            }
            if (this.getDiscordBot().getNetworkManager().isPlayerOnline(uuid, true)) {
                Player player = this.getDiscordBot().getNetworkManager().getPlayer(uuid);
                JsonEmbedBuilder jsonEmbedBuilder = JsonEmbedBuilder.fromJson(DCMessage.PLAYTIME_RESPONSE.getMessage());
                jsonEmbedBuilder.setTitle(jsonEmbedBuilder.getTitle().replace("%playername%", player.getName()));
                final List<MessageEmbed.Field> fields = new LinkedList<>();
                for (MessageEmbed.Field field : jsonEmbedBuilder.getFields()) {
                    String name = field.getName()
                            .replace("%playername%", player.getName())
                            .replace("%playtime%", TimeUtils.getTimeString(1, player.getPlaytime() / 1000))
                            .replace("%liveplaytime%", TimeUtils.getTimeString(player.getLanguage(), player.getPlaytime() / 1000));
                    String value = field.getValue()
                            .replace("%playername%", player.getName())
                            .replace("%playtime%", TimeUtils.getTimeString(1, player.getPlaytime() / 1000))
                            .replace("%liveplaytime%", TimeUtils.getTimeString(player.getLanguage(), player.getPlaytime() / 1000));
                    MessageEmbed.Field field1 = new MessageEmbed.Field(name, value, field.isInline());
                    fields.add(field1);
                }
                jsonEmbedBuilder.setFields(fields);
                Utils.sendChannelMessage(event.getTextChannel(),
                        jsonEmbedBuilder.build());
            } else {
                this.getDiscordBot().getScheduler().runAsync(() -> {
                    String[] result = getOfflinePlayerPlaytime(uuid);
                    JsonEmbedBuilder jsonEmbedBuilder = JsonEmbedBuilder.fromJson(DCMessage.PLAYTIME_RESPONSE.getMessage());
                    jsonEmbedBuilder.setTitle(jsonEmbedBuilder.getTitle().replace("%playername%", result[0]));
                    final List<MessageEmbed.Field> fields = new LinkedList<>();
                    for (MessageEmbed.Field field : jsonEmbedBuilder.getFields()) {
                        String name = field.getName()
                                .replace("%playername%", result[0])
                                .replace("%playtime%", TimeUtils.getTimeString(1, Long.parseLong(result[1]) / 1000))
                                .replace("%liveplaytime%", TimeUtils.getTimeString(1, Long.parseLong(result[1]) / 1000));
                        String value = field.getValue()
                                .replace("%playername%", result[0])
                                .replace("%playtime%", TimeUtils.getTimeString(1, Long.parseLong(result[1]) / 1000))
                                .replace("%liveplaytime%", TimeUtils.getTimeString(1, Long.parseLong(result[1]) / 1000));
                        MessageEmbed.Field field1 = new MessageEmbed.Field(name, value, field.isInline());
                        fields.add(field1);
                    }
                    jsonEmbedBuilder.setFields(fields);
                    Utils.sendChannelMessage(event.getTextChannel(),
                            jsonEmbedBuilder.build());
                }, false);
            }
        }
    }

    private String[] getOfflinePlayerPlaytime(UUID uuid) {
        String[] result = new String[]{};
        try (Connection connection = this.getDiscordBot().getMySQL().getConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT username, playtime FROM nm_players WHERE uuid=?")) {
            ps.setString(1, uuid.toString());
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    String username = rs.getString("username");
                    long playtime = rs.getLong("playtime");
                    result = new String[]{username, String.valueOf(playtime)};
                }
            }
        } catch (SQLException ex) {
            ex.printStackTrace();
        }
        return result;
    }

    private DiscordBot getDiscordBot() {
        return discordBot;
    }
}

File: src/main/java/nl/chimpgamer/networkmanager/extensions/discordbot/commands/discord/PlaytimeCommand.kt
package nl.chimpgamer.networkmanager.extensions.discordbot.commands.discord

import com.jagrosh.jdautilities.command.Command
import com.jagrosh.jdautilities.command.CommandEvent
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.ChannelType
import net.dv8tion.jda.api.entities.MessageEmbed
import nl.chimpgamer.networkmanager.api.utils.TimeUtils
import nl.chimpgamer.networkmanager.extensions.discordbot.DiscordBot
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.CommandSetting
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.DCMessage
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.JsonEmbedBuilder
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.Utils.sendChannelMessage
import java.sql.SQLException
import java.util.*

class PlaytimeCommand(private val discordBot: DiscordBot) : Command() {
    override fun execute(event: CommandEvent) {
        if (!event.isFromType(ChannelType.TEXT)) {
            return
        }
        if (!discordBot.commandSettings.getBoolean(CommandSetting.DISCORD_PLAYTIME_ENABLED)) {
            return
        }
        if (event.args.isEmpty()) {
            val uuid = discordBot.discordUserManager.getUuidByDiscordId(event.author.id)
            if (uuid == null) {
                discordBot.logger.warning("${event.author.name} tried to use the playtime command but is not registed!")
                return
            }
            if (discordBot.networkManager.isPlayerOnline(uuid, true)) {
                val player = discordBot.networkManager.getPlayer(uuid)
                val jsonEmbedBuilder = JsonEmbedBuilder.fromJson(discordBot.messages.getString(DCMessage.COMMAND_PLAYTIME_RESPONSE))
                jsonEmbedBuilder.title = jsonEmbedBuilder.title?.replace("%playername%", player.name)
                val fields: MutableList<MessageEmbed.Field> = LinkedList()
                for (field in jsonEmbedBuilder.fields) {
                    val name = field.name
                            ?.replace("%playername%", player.name)
                            ?.replace("%playtime%", TimeUtils.getTimeString(player.language, player.playtime / 1000))
                            ?.replace("%liveplaytime%", TimeUtils.getTimeString(player.language, player.playtime / 1000))
                    val value = field.value
                            ?.replace("%playername%", player.name)
                            ?.replace("%playtime%", TimeUtils.getTimeString(player.language, player.playtime / 1000))
                            ?.replace("%liveplaytime%", TimeUtils.getTimeString(player.language, player.playtime / 1000))
                    val field1 = MessageEmbed.Field(name, value, field.isInline)
                    fields.add(field1)
                }
                jsonEmbedBuilder.fields = fields
                sendChannelMessage(event.textChannel,
                        jsonEmbedBuilder.build())
            } else {
                discordBot.scheduler.runAsync(Runnable {
                    val result = getOfflinePlayerPlaytime(uuid)
                    val jsonEmbedBuilder = JsonEmbedBuilder.fromJson(discordBot.messages.getString(DCMessage.COMMAND_PLAYTIME_RESPONSE))
                    jsonEmbedBuilder.title = jsonEmbedBuilder.title?.replace("%playername%", result[0])
                    val fields: MutableList<MessageEmbed.Field> = LinkedList()
                    for (field in jsonEmbedBuilder.fields) {
                        val name = field.name
                                ?.replace("%playername%", result[0])
                                ?.replace("%playtime%", TimeUtils.getTimeString(1, result[1].toLong() / 1000))
                                ?.replace("%liveplaytime%", TimeUtils.getTimeString(1, result[1].toLong() / 1000))
                        val value = field.value
                                ?.replace("%playername%", result[0])
                                ?.replace("%playtime%", TimeUtils.getTimeString(1, result[1].toLong() / 1000))
                                ?.replace("%liveplaytime%", TimeUtils.getTimeString(1, result[1].toLong() / 1000))
                        val field1 = MessageEmbed.Field(name, value, field.isInline)
                        fields.add(field1)
                    }
                    jsonEmbedBuilder.fields = fields
                    sendChannelMessage(event.textChannel,
                            jsonEmbedBuilder.build())
                }, false)
            }
        }
    }

    private fun getOfflinePlayerPlaytime(uuid: UUID): Array<String> {
        var result = arrayOf<String>()
        try {
            discordBot.mySQL.connection.use { connection ->
                connection.prepareStatement("SELECT username, playtime FROM nm_players WHERE uuid=?").use { ps ->
                    ps.setString(1, uuid.toString())
                    ps.executeQuery().use { rs ->
                        if (rs.next()) {
                            val username = rs.getString("username")
                            val playtime = rs.getLong("playtime")
                            result = arrayOf(username, playtime.toString())
                        }
                    }
                }
            }
        } catch (ex: SQLException) {
            ex.printStackTrace()
        }
        return result
    }

    init {
        name = discordBot.commandSettings.getString(CommandSetting.DISCORD_PLAYTIME_COMMAND)
        cooldown = 3
        botPermissions = arrayOf(Permission.MESSAGE_WRITE)
        guildOnly = true
    }
}